### PR TITLE
Add datalog frontend via a proc macro

### PR DESCRIPTION
**Pull request by shadaj**

### CODE REVIEW

**Review by MingweiSamuel — COMMENTED** [review]
Nice!!

Couple comments before diving in:

1.
```
let (in_send, in_recv) = tokio::sync::mpsc::unbounded_channel::<(usize, usize)>();
...
.input in
```
Wouldn't it make sense to make the input `.input in_recv` so no `_send/_recv` convention needs to be enforced?

2. Snapshot testing will be fragile and the codegen will definitely change in the future so we should eventually find a better way to test

**Comment by shadaj** [comment]
Yeah perhaps we should snashot test the surface syntax or something higher level. Since we do have runtime tests, snapshots are mostly for my sanity of manually inspecting how changes affect different examples.

File: datalog_compiler/src/lib.rs [hunk 1]
```diff
@@ -0,0 +1,545 @@
...
+        .collect::<Vec<_>>();
+
+    let rules = program
+        .rules
+        .iter()
+        .filter_map(|decl| match decl {
+            Declaration::Rule(rule) => Some(rule.clone()),
+            _ => None,
+        })
+        .collect::<Vec<_>>();
```
Comment (MingweiSamuel) [inline datalog_compiler/src/lib.rs:47]: I'd prefer a single for loop here, or use https://docs.rs/separable/latest/separable/ if you wanna be fancy

File: datalog_compiler/src/lib.rs [hunk 2]
```diff
@@ -0,0 +1,545 @@
...
+                        },
+                        rhs: Box::new(Pipeline::Operator(Operator {
+                            path: parse_quote!(tee),
+                            paren_token: Paren::default(),
+                            args: Punctuated::new(),
+                        })),
+                    }),
+                },
+            ));
+        }
```
Comment (MingweiSamuel) [inline datalog_compiler/src/lib.rs:86]: Use [`syn::parse_quote!`](https://docs.rs/syn/latest/syn/macro.parse_quote.html) higher up? Like
```rust
{
    let name = &target_ident.name;
    parse_quote! {
        #name = merge() -> tee()
    }
}
```

File: datalog_compiler/src/lib.rs [hunk 3]
```diff
@@ -0,0 +1,545 @@
...
+                            span: Span::call_site(),
+                        },
+                    }),
+                },
+                rhs: Box::new(Pipeline::Name(syn::Ident::new(
+                    &target.name,
+                    Span::call_site(),
+                ))),
+            }),
+        ));
```
Comment (MingweiSamuel) [inline datalog_compiler/src/lib.rs:122]:
```rust
let my_merge_index_lit = syn::LitInt::new(format!("{}", my_merge_index), Span::call_site()); // or something
let name = &target.name;
parse_quote! {
    recv_stream(#target_ident) -> [#my_merge_index_lit] #name
}
```

File: datalog_compiler/src/lib.rs [hunk 4]
```diff
@@ -0,0 +1,545 @@
...
+    }
+
+    for target in outputs {
+        let tee_index = tee_counter.entry(target.name.clone()).or_insert(0);
+        let my_tee_index = *tee_index;
+        *tee_index += 1;
+
+        let out_send_ident = syn::Ident::new(&target.name, Span::call_site());
+
+        flat_graph.add_statement(hydroflow_lang::parse::HfStatement::Pipeline(
```
Comment (MingweiSamuel) [inline datalog_compiler/src/lib.rs:132]: `parse_quote!` ...

File: datalog_compiler/src/lib.rs (hunk 1 above)
Comment (MingweiSamuel) [reply datalog_compiler/src/lib.rs:47]: Also is the clone needed? Shared references should be ok for the lifetime of the fn?

File: datalog_compiler/src/lib.rs [hunk 5]
```diff
@@ -0,0 +1,545 @@
...
+            for (i, param) in source.fields.iter().enumerate() {
+                let entry = identifier_to_bindings
+                    .entry(param.clone())
+                    .or_insert_with(HashMap::new);
+                entry.insert(source_idx, i);
+            }
+        }
+
+        let mut sorted_identifiers = identifier_to_bindings.keys().cloned().collect::<Vec<_>>();
+        sorted_identifiers.sort_by_key(|ident| ident.name.clone());
```
Comment (MingweiSamuel) [inline datalog_compiler/src/lib.rs:178]: nit: I mean, maybe easier to just use BTreeMap?

File: datalog_compiler/src/lib.rs [hunk 6]
```diff
@@ -0,0 +1,545 @@
...
+                    parse_quote!(kv.1.#pair_idx)
+                };
+
+                let source_col_idx = syn::Index::from(*bindings.get(source).unwrap());
+
+                parse_quote!(#source_expr.#source_col_idx)
+            })
+            .collect::<Vec<syn::Expr>>();
+
+        // TODO(shadaj): this is all a hack, will fall apart when we have more than usize types
```
Comment (MingweiSamuel) [inline datalog_compiler/src/lib.rs:214]: :P

File: datalog_compiler/src/lib.rs [hunk 7]
```diff
@@ -0,0 +1,545 @@
...
+
+        let merge_index = merge_counter.entry(target.name.clone()).or_insert(0);
+        let my_merge_index = *merge_index;
+        *merge_index += 1;
+
+        let after_join = Pipeline::Link(PipelineLink {
+            lhs: Box::new(Pipeline::Operator(Operator {
+                path: parse_quote!(map),
+                paren_token: Paren::default(),
+                args: vec![after_join_map].iter().cloned::<syn::Expr>().collect(),
```
Comment (MingweiSamuel) [inline datalog_compiler/src/lib.rs:247]: `Punctuated::from_iter([after_join_map.clone()])` (though whole thing could be one parse_quote!)

File: datalog_compiler/src/lib.rs [hunk 8]
```diff
@@ -0,0 +1,545 @@
...
+
+        flat_graph.add_statement(hydroflow_lang::parse::HfStatement::Pipeline(
+            Pipeline::Link(PipelineLink {
+                lhs: Box::new(Pipeline::Operator(Operator {
+                    path: parse_quote!(recv_stream),
+                    paren_token: Paren::default(),
+                    args: vec![parse_quote!(#target_ident)]
+                        .iter()
+                        .cloned::<syn::Expr>()
+                        .collect(),
```
Comment (MingweiSamuel) [inline datalog_compiler/src/lib.rs:104]: cleaner: `Punctuated::from_iter([parse_quote!(#target_ident)])`

Also clone is unnecessary?

File: datalog_compiler/src/lib.rs [hunk 9]
```diff
@@ -0,0 +1,545 @@
...
+                    arrow: Token![->](Span::call_site()),
+                    dst: None,
+                },
+                rhs: Box::new(Pipeline::Operator(Operator {
+                    path: parse_quote!(for_each),
+                    paren_token: Paren::default(),
+                    args: vec![parse_quote!(|v| #out_send_ident.send(v).unwrap())]
+                        .iter()
+                        .cloned::<syn::Expr>()
+                        .collect(),
```
Comment (MingweiSamuel) [inline datalog_compiler/src/lib.rs:155]: `Punctuated::from_iter([parse_quote!(|v| #out_send_ident.send(v).unwrap())])`

File: datalog_compiler/src/lib.rs [hunk 10]
```diff
@@ -0,0 +1,545 @@
...
+                        Pipeline::Name(join_node.clone())
+                    } else {
+                        Pipeline::Link(PipelineLink {
+                            lhs: Box::new(Pipeline::Operator(Operator {
+                                path: parse_quote!(map),
+                                paren_token: Paren::default(),
+                                args: vec![parse_quote!(|v: (#(#source_data_types, )*)| ((#(#hash_keys, )*), v))]
+                                    .iter()
+                                    .cloned::<syn::Expr>()
+                                    .collect(),
```
Comment (MingweiSamuel) [inline datalog_compiler/src/lib.rs:339]: `Punctuated::from_iter([parse_quote!(|v: (#(#source_data_types, )*)| ((#(#hash_keys, )*), v))])`

File: datalog_compiler/src/lib.rs [hunk 11]
```diff
@@ -0,0 +1,545 @@
...
+                src: None,
+                arrow: Token![->](Span::call_site()),
+                dst: Some(Indexing {
+                    bracket_token: syn::token::Bracket::default(),
+                    index: IndexInt {
+                        value: my_merge_index,
+                        span: Span::call_site(),
+                    },
+                }),
+            },
```
Comment (MingweiSamuel) [inline datalog_compiler/src/lib.rs:259]: one `parse_quote!`?

File: datalog_compiler/src/lib.rs [hunk 12]
```diff
@@ -0,0 +1,545 @@
...
+                    paren_token: Paren::default(),
+                    args: Punctuated::new(),
+                })),
+                connector: ArrowConnector {
+                    src: None,
+                    arrow: Token![->](Span::call_site()),
+                    dst: None,
+                },
+                rhs: Box::new(after_join),
+            })
```
Comment (MingweiSamuel) [inline datalog_compiler/src/lib.rs:278]: one `parse_quote!`?

File: datalog_compiler/src/lib.rs [hunk 13]
```diff
@@ -0,0 +1,545 @@
...
+        };
+
+        flat_graph.add_statement(hydroflow_lang::parse::HfStatement::Named(
+            NamedHfStatement {
+                name: join_node.clone(),
+                equals: Token![=](Span::call_site()),
+                pipeline: join_and_map,
+            },
+        ));
```
Comment (MingweiSamuel) [inline datalog_compiler/src/lib.rs:287]: one `parse_quote!`?

File: datalog_compiler/src/lib.rs [hunk 14]
```diff
@@ -0,0 +1,545 @@
...
+                .iter()
+                .map(|(ident, bindings)| {
+                    if let Some(idx) = bindings.get(&source_i) {
+                        let idx_ident = syn::Index::from(*idx);
+                        parse_quote!(v.#idx_ident)
+                    } else {
+                        panic!("Could not find key that is being joined on: {:?}", ident);
+                    }
+                })
+                .collect::<Vec<_>>();
```
Comment (MingweiSamuel) [inline datalog_compiler/src/lib.rs:300]: nit: `.collect();`

File: datalog_compiler/src/lib.rs [hunk 15]
```diff
@@ -0,0 +1,545 @@
...
+                                        span: Span::call_site(),
+                                    },
+                                }),
+                            },
+                            rhs: Box::new(Pipeline::Name(join_node.clone())),
+                        })
+                    }),
+                }),
+            ));
+        }
```
Comment (MingweiSamuel) [inline datalog_compiler/src/lib.rs:357]: one `parse_quote!`?

**Review by MingweiSamuel — APPROVED** [review]
Would be good to have some high-level comments through the big function. Or split into a few smaller functions if that makes sense.

got a bit nit-ey in the review :P

File: datalog_compiler/src/lib.rs [hunk 16]
```diff
@@ -0,0 +1,546 @@
...
+use syn::{parse_quote, punctuated::Punctuated, token::Paren, Token};
+
+mod grammar;
+
+use grammar::datalog::*;
+
+fn gen_datalog_program(literal: proc_macro2::Literal, root: TokenStream) -> syn::Stmt {
+    let str_node: syn::LitStr = parse_quote!(#literal);
+    let actual_str = str_node.value();
+    let program: Program = grammar::datalog::parse(&actual_str).unwrap();
```
[the remark this answers is not in the document]
Comment (shadaj) [reply datalog_compiler/src/lib.rs:20]: Yep, Rust Sitter actually gives us error spans but need to write the logic that spits out diagnostics: #181

**Pull request merged**